[Pipeline] Add retimeable pipeline operation

mortbopet opened the pull request:
The `pipeline.rtp` operation represents a retimeable pipeline.
The pipeline contains a single block representing a graph region. Pipeline stages are represented by `pipeline.rt.register` operations. Semantics of values crossing register boundaries are defined by lowering passes.
The pipeline representation is centered around providing latency insensitive values (valid signals between stages). Such signals can either be removed (in the feedforward, statically scheduled case), used to control stalling (feedback, statically scheduled case) or in conjunction with handshake signals for dynamically scheduled pipelines.

Alongside this, some old, unused code (StandardToStaticLogic) is pruned to remove a maintenance burden associated with this change.

A typical flow would go like this:

An untimed datapath is defined:
```mlir
pipeline.rtp(%in0 : i32, %in1 : i32,  %go : i1) -> (i32) {
    ^bb0:(%arg0 : i32, %arg1: i32, %go : i1):
    %add0 = comb.add %arg0, %arg1 : i32
    %add1 = comb.add %add0, %arg0 : i32
    %add2 = comb.add %add1, %arg1 : i32
    pipeline.rtp.return %add2 valid %go : i32
}
```
The datapath is scheduled (future pass). At this point, stage separating registers can be inserted at arbitrary points in the datapath. In future commits, delays will always be specified relative to the op that produces it - either by `pipeline.delay %0 by N` operations that can delay an arbitrary SSA value across the next `N` stages, or annotating that some values take multiple cycles to produce (pipelined operators).
```mlir
pipeline.rtp(%in0 : i32, %in1 : i32, %go : i1) -> (i32) {
    ^bb0:(%arg0 : i32, %arg1: i32, %go : i1):
    %add0 = comb.add %arg0, %arg1 : i32

    %s0_valid = pipeline.stage when %go
    %add1 = comb.add %add0, %arg0 : i32

    %s1_valid = pipeline.stage when %s0_valid
    %add2 = comb.add %add1, %arg1 : i32

    pipeline.rtp.return %add2 valid %s1_valid : i32
}
```

Stage-crossing dependencies are made explicit through registers (future pass). At this point, `pipeline.stage` operations are illegal, and the pipeline has been timed.
```mlir
pipeline.rtp(%in0 : i32, %in1 : i32,  %go : i1) -> (i32) {
    ^bb0:(%arg0 : i32, %arg1: i32,  %go : i1):
    %add0 = comb.add %arg0, %arg1 : i32

    %s0_valid, %add0_r = pipeline.stage.register when %go regs (%add0: i32)
    %add1 = comb.add %add0_r, %arg0 : i32

    %s1_valid, %add1_r = pipeline.stage.register when %s0_valid regs (%add1: i32)
    %add2 = comb.add %add1_r, %arg1 : i32

    pipeline.rtp.return %add2 valid %s1_valid : i32
}
```

This representation can then be lowered to statically or dynamically scheduled pipelines.

teqdruid commented:
The rename is all mixed in so the diff is impossible to read. will review once the rename is merged.

One comment: I hate the name "retimeable" -- it's a misnomer. What are you "timing" again? Additionally, it's really just a blob of logic which can optionally be pipelined OR run combinationally. I would prefer a much more generic name -- like 'dataflow' or something like that -- which implies what it contains and means rather than the intent of the transformation.

mortbopet commented:
> "retimeable" -- it's a misnomer. What are you "timing" again?

Naming is hard. Retimeable seemed like a good fit since the `pipeline.stage` operations can be freely added, removed and moved around within the body of the pipeline. Obviously it aliases with circuit retiming. I would be OK with `pipeline.datapath` - @mikeurbach thoughts?

> Additionally, it's really just a blob of logic which can optionally be pipelined OR run combinationally

I don't think this is the sole intent of this op (though it might be, if we find that restriction to be the right thing to do). In the slides which @mikeurbach will present, it shows how this representation is also used to describe dataflows with delayed values across to-be-decided stages, backedges etc. - as such, a higher level of abstraction than just feed-forward combinational logic (which will be the immediate use-case that PyCDE will target).

mortbopet commented:
This has now been rebased on top of the dialect name change.

jopperm commented:
`%g1` should be `%s0_valid`, right?

jopperm commented:
Intriguing design! One question though: I assume the "middle" IR (w/ stages, w/o regs) is intended to conveniently transform the pipeline by moving around ops between stages. Given that the pipeline's body is a graph region, isn't it problematic then that the association of ops to stages is encoded by the relative location inside the block?

mortbopet commented:
> Given that the pipeline's body is a graph region, isn't it problematic then that the association of ops to stages is encoded by the relative location inside the block?

It _would_ be if a region being a graph region would imply a non-deterministic ordering between the order of operations in printed IR and in in-memory. While this property is not formalized, i think it is a valid design point (and safe - i don't assume the underlying data structures will change any time soon).

The representation seems well suited for a pipelining representation - in a sense, it's a `GraphRegion` which implicitly states something about the topology of the graph.

mikeurbach commented:
> the slides which @mikeurbach will present

For the record: https://docs.google.com/presentation/d/1PXgyVRmOMqf2PkshGgodLfwFiJAqh00E_yga2IAEvAM/edit#slide=id.gf491adc922_0_10. Still looking for an ODM slot to dicsuss, but some of what Morten proposes here is in slides 6-8.

> in a sense, it's a GraphRegion which implicitly states something about the topology of the graph through the reliance on operation ordering.

That sounds reasonable to me, it's been done before: https://circt.llvm.org/docs/Dialects/SV/#svordered-circtsvorderedoutputop

mortbopet commented:
@mikeurbach @teqdruid any further thoughts on this? Eager to get it into tree so i can post some of the related passes.

mikeurbach commented:
Sorry I haven't had time to give this a proper review, but will try to get on that early this week.

jopperm commented on include/circt/Dialect/Pipeline/Pipeline.td:
```diff
@@ -14,55 +14,133 @@
+  }];
+
+  let extraClassDeclaration = [{
+    // Implement RegionKindInterface.
+    static RegionKind getRegionKind(unsigned index) { return RegionKind::Graph;}
```
Nit: Use the `HasOnlyGraphRegion` trait.

jopperm commented:
> > "retimeable" -- it's a misnomer. What are you "timing" again?
> 
> Naming is hard. Retimeable seemed like a good fit since the `pipeline.stage` operations can be freely added, removed and moved around within the body of the pipeline.

Actually, the "retimable" moniker made sense to me, for the reasons Morten listed 😉 Anyways, `datapath` would also SGTM. Or could this just be *the* `PipelineOp` in the `pipeline` dialect (like `func.FuncOp`)?

mortbopet commented:
> Actually, the "retimable" moniker made sense to me, for the reasons Morten listed 😉 Anyways, datapath would also SGTM. Or could this just be _the_ PipelineOp in the pipeline dialect (like func.FuncOp)?

I have no issues with this - it seems like this op can represent most (all) semantics of a pipeline while, and I'd expect `PipelineWhileOp` to be able to lower to it. So this seems natural - will change.

teqdruid commented:
> > Actually, the "retimable" moniker made sense to me, for the reasons Morten listed 😉 Anyways, datapath would also SGTM. Or could this just be _the_ PipelineOp in the pipeline dialect (like func.FuncOp)?
> 
> I have no issues with this - it seems like this op can represent most (all) semantics of a pipeline while, and I'd expect `PipelineWhileOp` to be able to lower to it. So this seems natural - will change.

No objections here!

teqdruid reviewed (commented):
This looks good! You're missing the error tests (which you're usually quite good about), so I don't think you're done.

Can you break out and directly commit the PipelineWhile rename?

mortbopet commented on include/circt/Dialect/Pipeline/Pipeline.td:
```diff
@@ -14,21 +14,145 @@
+      ^bb0:(%arg0 : i32, %arg1: i32):
+        %add0 = comb.add %arg0, %arg1 : i32
+
+        %s0_valid, %add0_r = pipeline.stage when %go regs (%add0: i32)
```
> Are the registers detected automatically?

Yes - locally, i have a `pipeline-make-regs-explicit` pass which performs a dataflow analysis and mutates the `pipeline.stage` operations accordingly (def-use references across stages).

> Are the registers ... required?

No - this is by design. The lowering pass is fairly dumb in that it pretty much just convers the `pipeline.stage` operation to `seq.compreg` operations. Any dataflow analysis and mutation of stage ops should have done beforehand. The intention here is to keep things separated, that is, analyzing and inferring registers ("hard"-ish) and lowering to HW (easy).

teqdruid replied:
I must be missing something... If a cross stage value isn't listed, won't the resulting circt be wrong? Of at the very least have an II >= 2 (since that value could be a multi-cycle signal).

mortbopet replied:
> If a cross stage value isn't listed, won't the resulting circt be wrong? Of at the very least have an II >= 2 (since that value could be a multi-cycle signal).

We might have differing view of at what level this pipeline models the operations. At the moment, it strictly serves as a notion of adding registers to evenly break combinational paths in a dataflow graph. There is no notion of initiation interval or multi-cycle signals.

A pipeline
```mlir
pipeline.pipeline ... {
    %0 = ...
    %1 = ...
    %valid = pipeline.stage when %go 
    %2 = comb.add %0, %1
}
```
lowers to
```mlir
%0 = ...
%1 = ...
%valid = comb.reg %go ...
%2 = comb.add %0, %1
```

whereas if the stage register materialization pass has been run on the pipeline:
```mlir
pipeline.pipeline ... {
    %0 = ...
    %1 = ...
    %valid, %0_r, %1_r = pipeline.stage when %go regs %0 %1
    %2 = comb.add %0_r, %1_r
}
```

that lowers to
```mlir
%0 = ...
%1 = ...
%valid = comb.reg %go ...
%0_r = comb.reg %0 ...
%1_r = comb.reg %1 ...
%2 = comb.add %0_r, %1_r
```

teqdruid replied:
> We might have differing view of at what level this pipeline models the operations. At the moment, it strictly serves as a notion of adding registers to evenly break combinational paths in a dataflow graph. There is no notion of initiation interval or multi-cycle signals.

In the presence of multi-cycle paths, every data pipeline has a notion of II since with a multi-cycle path, one can logically only show new (meaningful/valid) data every N cycles (wherein N is the length of the multi-cycle path), yes?

The term 'stage' implies -- to me -- something like a 'barrier' wherein all of the signals moving into the next stage must be registered. So in your first example, assuming that `%0` and `%1` are combinationally driven by inputs and `%2` and `%valid` combinationally drives an output, then `%valid` will be output a cycle 'late' by my mental model of a pipeline.

mortbopet replied:
> In the presence of multi-cycle paths, every data pipeline has a notion of II since with a multi-cycle path, one can logically only show new (meaningful/valid) data every N cycles (wherein N is the length of the multi-cycle path), yes?

I don't disagree with you here, I just don't want to move the goalpost for what this PR and abstraction intends to capture in its initial version. These are important things to capture but I'd like to get the basic support in tree and working, before tackling more advanced cases. I intend upcoming commits to add support for e.g. `pipeline.delay` which delays a signal across the next N stages, and we could also add a rudimentary `cycles` attribute to arbitrary SSA values to inform register materialization that a path is multi-cycle.

> The term 'stage' implies -- to me -- something like a 'barrier' wherein all of the signals moving into the next stage must be registered. So in your first example, assuming that %0 and %1 are combinationally driven by inputs and %2 and %valid combinationally drives an output, then %valid will be output a cycle 'late' by my mental model of a pipeline.

The intention of the design is to have an abstraction that easily facilitates insertion of new stages/moving stages around (hence the old 'retimeable' name). This is why hardering/materialization of registers is not a required part of the stage operation but instead a pass late in the pass pipeline.  
I see this abstraction as having different modes - one before, and one after register materialization. Your barrier notion matches my view of the abstraction under the assumption that register materialization is run before lowering.

mortbopet replied:
(Discussed offline)
Added `pipeline.stage.register` and removed the `regs` capability from `pipeline.stage`. These two operations cannot exist together, with `pipeline.stage` lowering to `pipeline.stage.registers` (see updated example at top).

mortbopet commented:
> You're missing the error tests

Woops - added!

> Can you break out and directly commit the PipelineWhile rename?

5fc5b1d3726edfee614482727f6044273dd1a6b7

teqdruid reviewed (approved):
LGTM! thanks for taking the time to understand my concerns.
